1.20 crash

### Expected behavior

no crash like folia-1.19.4

### Observed/Actual behavior

https://pastes.dev/GdgFr21jMO

### Steps/models to reproduce

just tp around

### Plugin and Datapack List

> plugins
[09:50:18 INFO]: Server Plugins (19):
[09:50:18 INFO]: Bukkit Plugins:
[09:50:18 INFO]:  - AnCDK, ArmorStandEditor, Chunky, ColoredAnvils, CoreProtect, CreeperConfetti, DeathMessage, Essentials, EssentialsChat, GSit
[09:50:18 INFO]:  InvSeePlusPlus, Kaiivoid, LuckPerms, NoFlightInEnd, PistonMOTD, SkullPlugin, TAB, Vault, voicechat

### Folia version

> version
[09:50:52 INFO]: Checking version, please wait...
[09:50:53 INFO]: This server is running Folia version git-Folia-"9dcd6c2" (MC: 1.20) (Implementing API version 1.20-R0.1-SNAPSHOT) (Git: 9dcd6c2 on dev/1.20)
Error obtaining version information
Previous version: git-Folia-"3fde784" (MC: 1.20)

### Other

_No response_

Thread Region Scheduler Thread #4 failed main thread check: Cannot remove entity off-main
java.lang.Throwable: null
	at io.papermc.paper.util.TickThread.ensureTickThread(TickThread.java:78) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at net.minecraft.world.entity.Entity.setRemoved(Entity.java:5477) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at io.papermc.paper.world.ChunkEntitySlices.unload(ChunkEntitySlices.java:122) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at io.papermc.paper.chunk.system.scheduling.NewChunkHolder.unloadStage2(NewChunkHolder.java:850) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at io.papermc.paper.chunk.system.scheduling.ChunkHolderManager.processUnloads(ChunkHolderManager.java:1306) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at net.minecraft.server.level.ChunkMap.processUnloads(ChunkMap.java:521) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at net.minecraft.server.level.ChunkMap.tick(ChunkMap.java:509) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at net.minecraft.server.level.ServerChunkCache.tick(ServerChunkCache.java:455) ~[?:?]
	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:793) ~[?:?]
	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1678) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:463) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1535) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at io.papermc.paper.threadedregions.TickRegions$ConcreteRegionTickHandle.tickRegion(TickRegions.java:360) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at io.papermc.paper.threadedregions.TickRegionScheduler$RegionScheduleHandle.runTick(TickRegionScheduler.java:385) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at ca.spottedleaf.concurrentutil.scheduler.SchedulerThreadPool$TickThreadRunner.run(SchedulerThreadPool.java:525) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
[16:14:09] [Region Scheduler Thread #4/ERROR]: [io.papermc.paper.threadedregions.TickRegionScheduler] Region #4 centered at chunk [549, 2164] in world 'world' failed to tick:
net.minecraft.ReportedException: Exception ticking world
	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1697) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:463) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1535) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at io.papermc.paper.threadedregions.TickRegions$ConcreteRegionTickHandle.tickRegion(TickRegions.java:360) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at io.papermc.paper.threadedregions.TickRegionScheduler$RegionScheduleHandle.runTick(TickRegionScheduler.java:385) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at ca.spottedleaf.concurrentutil.scheduler.SchedulerThreadPool$TickThreadRunner.run(SchedulerThreadPool.java:525) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.IllegalStateException: Cannot remove entity off-main
	at io.papermc.paper.util.TickThread.ensureTickThread(TickThread.java:79) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at net.minecraft.world.entity.Entity.setRemoved(Entity.java:5477) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at io.papermc.paper.world.ChunkEntitySlices.unload(ChunkEntitySlices.java:122) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at io.papermc.paper.chunk.system.scheduling.NewChunkHolder.unloadStage2(NewChunkHolder.java:850) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at io.papermc.paper.chunk.system.scheduling.ChunkHolderManager.processUnloads(ChunkHolderManager.java:1306) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	at net.minecraft.server.level.ChunkTracker.processUnloads(ChunkTracker.java:521) ~[?:?]
	at net.minecraft.server.level.ChunkTracker.tick(ChunkTracker.java:509) ~[?:?]
	at net.minecraft.server.level.ServerChunkCache.tick(ServerChunkCache.java:455) ~[?:?]
	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:793) ~[?:?]
	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1678) ~[kaiiju-1.20.jar:git-Kaiiju-"21321ab"]
	... 6 more
[16:14:09] [Region shutdown thread/INFO]: [RegionShutdownThread] Awaiting scheduler termination for 60s
[16:14:09] [Region shutdown thread/INFO]: [RegionShutdownThread] Scheduler halted
[16:14:09] [Region shutdown thread/INFO]: Stopping server

[09:43:15] [Region Scheduler Thread #1/ERROR]: Thread Region Scheduler Thread #1 failed main thread check: Cannot unload entity section off-main
java.lang.Throwable: null
	at io.papermc.paper.util.TickThread.ensureTickThread(TickThread.java:71) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at io.papermc.paper.chunk.system.entity.EntityLookup.entitySectionUnload(EntityLookup.java:702) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at io.papermc.paper.chunk.system.scheduling.NewChunkHolder.unloadStage2(NewChunkHolder.java:859) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at io.papermc.paper.chunk.system.scheduling.ChunkHolderManager.processUnloads(ChunkHolderManager.java:1306) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at net.minecraft.server.level.ChunkMap.processUnloads(ChunkMap.java:521) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at net.minecraft.server.level.ChunkMap.tick(ChunkMap.java:509) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at net.minecraft.server.level.ServerChunkCache.tick(ServerChunkCache.java:455) ~[?:?]
	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:793) ~[?:?]
	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1678) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:447) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1535) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at io.papermc.paper.threadedregions.TickRegions$ConcreteRegionTickHandle.tickRegion(TickRegions.java:360) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at io.papermc.paper.threadedregions.TickRegionScheduler$RegionScheduleHandle.runTick(TickRegionScheduler.java:385) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at ca.spottedleaf.concurrentutil.scheduler.SchedulerThreadPool$TickThreadRunner.run(SchedulerThreadPool.java:525) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at java.lang.Thread.run(Thread.java:1623) ~[?:?]
[09:43:16] [Region Scheduler Thread #1/ERROR]: [io.papermc.paper.threadedregions.TickRegionScheduler] Region #1 centered at chunk [0, 0] in world 'world' failed to tick:
net.minecraft.ReportedException: Exception ticking world
	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1697) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:447) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1535) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at io.papermc.paper.threadedregions.TickRegions$ConcreteRegionTickHandle.tickRegion(TickRegions.java:360) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at io.papermc.paper.threadedregions.TickRegionScheduler$RegionScheduleHandle.runTick(TickRegionScheduler.java:385) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at ca.spottedleaf.concurrentutil.scheduler.SchedulerThreadPool$TickThreadRunner.run(SchedulerThreadPool.java:525) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at java.lang.Thread.run(Thread.java:1623) ~[?:?]
Caused by: java.lang.IllegalStateException: Cannot unload entity section off-main
	at io.papermc.paper.util.TickThread.ensureTickThread(TickThread.java:72) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at io.papermc.paper.chunk.system.entity.EntityLookup.entitySectionUnload(EntityLookup.java:702) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at io.papermc.paper.chunk.system.scheduling.NewChunkHolder.unloadStage2(NewChunkHolder.java:859) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at io.papermc.paper.chunk.system.scheduling.ChunkHolderManager.processUnloads(ChunkHolderManager.java:1306) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	at net.minecraft.server.level.ChunkTracker.processUnloads(ChunkTracker.java:521) ~[?:?]
	at net.minecraft.server.level.ChunkTracker.tick(ChunkTracker.java:509) ~[?:?]
	at net.minecraft.server.level.ServerChunkCache.tick(ServerChunkCache.java:455) ~[?:?]
	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:793) ~[?:?]
	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1678) ~[folia-1.20.jar:git-Folia-"9b2ffd0"]
	... 6 more
[09:43:16] [Region shutdown thread/INFO]: [RegionShutdownThread] Awaiting scheduler termination for 60s
[09:43:16] [Region shutdown thread/INFO]: [RegionShutdownThread] Scheduler halted
[09:43:16] [Region shutdown thread/INFO]: Stopping server
[09:43:16] [Region shutdown thread/INFO]: [motdgg] Disabling motdgg v1.3.4
[09:43:16] [Region shutdown thread/INFO]: [BlueMap] Disabling BlueMap v3.14
[09:43:16] [Region shutdown thread/INFO]: [BlueMap] Stopping...
[09:43:16] [Region shutdown thread/INFO]: [BlueMap] Saved and stopped!
[09:43:16] [Region shutdown thread/INFO]: [RegionShutdownThread] Halting chunk systems
[09:43:16] [Region shutdown thread/INFO]: [ChunkHolderManager] Waiting 60s for chunk system to halt for world 'world'
[09:43:16] [Region shutdown thread/INFO]: [ChunkHolderManager] Halted chunk system for world 'world'
[09:43:16] [Region shutdown thread/INFO]: [ChunkHolderManager] Waiting 60s for chunk system to halt for world 'world_nether'
[09:43:16] [Region shutdown thread/INFO]: [ChunkHolderManager] Halted chunk system for world 'world_nether'
[09:43:16] [Region shutdown thread/INFO]: [ChunkHolderManager] Waiting 60s for chunk system to halt for world 'world_the_end'
[09:43:16] [Region shutdown thread/INFO]: [ChunkHolderManager] Halted chunk system for world 'world_the_end'
[09:43:16] [Region shutdown thread/INFO]: [RegionShutdownThread] Halted chunk systems
[09:43:16] [Region shutdown thread/INFO]: [RegionShutdownThread] Saving chunks around region around chunk [0, 0] in world 'world'
[09:43:16] [Region shutdown thread/INFO]: [ChunkHolderManager] Saved 18 block chunks, 99 entity chunks, 1 poi chunks in world 'world' in 0.09s
[09:43:16] [Region shutdown thread/INFO]: [RegionShutdownThread] Saving chunks around region around chunk [-22, 186] in world 'world'
[09:43:18] [Region shutdown thread/INFO]: [ChunkHolderManager] Saved 1226 block chunks, 1226 entity chunks, 0 poi chunks in world 'world' in 1.77s
[09:43:18] [Region shutdown thread/INFO]: [RegionShutdownThread] Saving chunks around region around chunk [0, 0] in world 'world_nether'
[09:43:18] [Region shutdown thread/INFO]: [ChunkHolderManager] Saved 158 block chunks, 90 entity chunks, 0 poi chunks in world 'world_nether' in 0.20s
[09:43:18] [Region shutdown thread/INFO]: [RegionShutdownThread] Saving chunks around region around chunk [0, 0] in world 'world_the_end'
[09:43:18] [Region shutdown thread/INFO]: [ChunkHolderManager] Saved 1 block chunks, 40 entity chunks, 0 poi chunks in world 'world_the_end' in 0.01s
[09:43:18] [Region shutdown thread/INFO]: [RegionShutdownThread] Saving players
[09:43:18] [Region shutdown thread/INFO]: Flushing Chunk IO
[09:43:18] [Region shutdown thread/INFO]: Closing Thread Pool
[09:43:18] [Region shutdown thread/INFO]: Closing Server

--- a/patches/server/0003-Threaded-Regions.patch
+++ b/patches/server/0003-Threaded-Regions.patch
@@ -1838,7 +1838,7 @@ index 82ccaf612548a7dbab7e5aeffb6eb8db84367477..b9095f559472dd92375ea719886913f6
      }
  
 diff --git a/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkHolderManager.java b/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkHolderManager.java
-index abd0217cf0bff183c8e262edc173a53403797c1a..33524deb5e1eda5be53e5b426c88f5837eb3e512 100644
+index abd0217cf0bff183c8e262edc173a53403797c1a..d496ea6a583f71ddfc17ada1424c8c7a026fdf4d 100644
 --- a/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkHolderManager.java
 +++ b/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkHolderManager.java
 @@ -53,6 +53,14 @@ import java.util.concurrent.atomic.AtomicReference;
@@ -2192,6 +2192,15 @@ index abd0217cf0bff183c8e262edc173a53403797c1a..33524deb5e1eda5be53e5b426c88f583
          synchronized (this.chunkHolders) {
              this.chunkHolders.remove(CoordinateUtils.getChunkKey(holder.chunkX, holder.chunkZ));
          }
+@@ -1058,7 +1195,7 @@ public final class ChunkHolderManager {
+             throw new IllegalStateException("Cannot unload chunks recursively");
+         }
+         final int sectionShift = this.unloadQueue.coordinateShift; // sectionShift <= lock shift
+-        final List<ChunkQueue.SectionToUnload> unloadSectionsForRegion = this.unloadQueue.retrieveForAllRegions();
++        final List<ChunkQueue.SectionToUnload> unloadSectionsForRegion = this.unloadQueue.retrieveForCurrentRegion(); // Folia - threaded regions
+         int unloadCountTentative = 0;
+         for (final ChunkQueue.SectionToUnload sectionRef : unloadSectionsForRegion) {
+             final ChunkQueue.UnloadSection section
 @@ -1371,7 +1508,13 @@ public final class ChunkHolderManager {
  
      // only call on tick thread
@@ -2227,6 +2236,59 @@ index abd0217cf0bff183c8e262edc173a53403797c1a..33524deb5e1eda5be53e5b426c88f583
  
          final JsonArray unloadQueue = new JsonArray();
          ret.add("unload_queue", unloadQueue);
+diff --git a/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkQueue.java b/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkQueue.java
+index 4cc1b3ba6d093a9683dbd8b7fe76106ae391e019..47bbd5de5849bef5392e3783322a19de5b351beb 100644
+--- a/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkQueue.java
++++ b/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkQueue.java
+@@ -1,5 +1,8 @@
+ package io.papermc.paper.chunk.system.scheduling;
+ 
++import io.papermc.paper.threadedregions.ThreadedRegionizer;
++import io.papermc.paper.threadedregions.TickRegionScheduler;
++import io.papermc.paper.threadedregions.TickRegions;
+ import it.unimi.dsi.fastutil.HashCommon;
+ import it.unimi.dsi.fastutil.longs.LongLinkedOpenHashSet;
+ import java.util.ArrayList;
+@@ -45,6 +48,39 @@ public final class ChunkQueue {
+         return ret;
+     }
+ 
++    // Folia start - threaded regions
++    public List<SectionToUnload> retrieveForCurrentRegion() {
++        final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region =
++            TickRegionScheduler.getCurrentRegion();
++        final ThreadedRegionizer<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> regionizer = region.regioniser;
++        final int shift = this.coordinateShift;
++
++        final List<SectionToUnload> ret = new ArrayList<>();
++
++        for (final Map.Entry<Coordinate, UnloadSection> entry : this.unloadSections.entrySet()) {
++            final Coordinate coord = entry.getKey();
++            final long key = coord.key;
++            final UnloadSection section = entry.getValue();
++            final int sectionX = Coordinate.x(key);
++            final int sectionZ = Coordinate.z(key);
++            final int chunkX = sectionX << shift;
++            final int chunkZ = sectionZ << shift;
++
++            if (regionizer.getRegionAtUnsynchronised(chunkX, chunkZ) != region) {
++                continue;
++            }
++
++            ret.add(new SectionToUnload(sectionX, sectionZ, coord, section.order, section.chunks.size()));
++        }
++
++        ret.sort((final SectionToUnload s1, final SectionToUnload s2) -> {
++            return Long.compare(s1.order, s2.order);
++        });
++
++        return ret;
++    }
++    // Folia end - threaded regions
++
+     public UnloadSection getSectionUnsynchronized(final int sectionX, final int sectionZ) {
+         final Coordinate coordinate = new Coordinate(Coordinate.key(sectionX, sectionZ));
+         return this.unloadSections.get(coordinate);
 diff --git a/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkTaskScheduler.java b/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkTaskScheduler.java
 index f975cb93716e137d973ff2f9011acdbef58859a2..b84bfc9513a13e365f2bd471b3c77058638d1384 100644
 --- a/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkTaskScheduler.java

